[방탈출 예약 대기 STEP2] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
안녕하세요 로키 😄 
이번 리뷰도 잘 부탁드립니다 🫶 

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/4400

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- [x]  한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- [x]  else 예약어를 쓰지 않는다.
- [x]  모든 원시 값과 문자열을 포장한다.
- [ ]  일급 컬렉션을 쓴다.
- [x]  한 줄에 점을 하나만 찍는다.
- [x]  줄여 쓰지 않는다(축약 금지).
- [x]  모든 엔티티를 작게 유지한다.
- [ ]  3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [x]  getter/setter/프로퍼티를 쓰지 않는다.


## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

코멘트로 남기겠습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

yesjuhee committed:
feat: 예약 대기 클라이언트 코드 추가

yesjuhee committed:
refactor: 안쓰는 API 삭제

yesjuhee committed:
docs: 예약 대기 요구사항 추가

yesjuhee committed:
docs: 예약 대기 응답 형식 수정

yesjuhee committed:
feat: 예약 대기 생성 API 구현

yesjuhee committed:
docs: 내 예약 조회 응답 형식 수정

yesjuhee committed:
refactor: 내 예약 조회 시 대기 순위 표시 기능 추가

- 나의 예약 목록에 예약 대기 데이터 및 순위 표시 로직 추가
- `WaitingWithRank` 도메인 클래스와 관련 쿼리 작성
- 예약 상태를 나타내는 `ReservationStatus` 및 응답 객체 추가
- 프론트엔드에서 예약 대기 상태와 순위를 출력하도록 수정
- 초기 데이터에 예약 대기 관련 데이터 추가

yesjuhee committed:
docs: 예약 대기 취소 API 엔드포인트 추가 및 문서 시나리오 보강

- 예약 대기 취소 API 호출을 위한 엔드포인트 추가 (`/waitings/${id}`)
- 예약 및 예약 대기에 대한 API 문서 시나리오 추가

yesjuhee committed:
feat: 예약 대기 삭제 기능 추가

- 사용자가 자신의 예약 대기 내역을 삭제할 수 있는 기능을 추가
- DELETE /waitings/{id} API 구현 및 관련 권한 체크 로직 적용
- 테스트 코드에 예약 대기 삭제 관련 케이스 추가
- 테스트 데이터 SQL 수정 및 추가

yesjuhee committed:
fix: 내 예약 목록 반환 조회 오류 수정

- `waiting`의 `id`값을 반환하도록 수정

yesjuhee committed:
feat: 예약 대기 관리 기능 관련 클라이언트 코드 추가

- 예약 대기 목록 페이지(waiting.html) 및 관련 JS(waiting.js) 파일 추가
- 어드민 네비게이션 메뉴에 예약 대기 관리 링크 추가

yesjuhee committed:
docs: 어드민 예약 대기 목록 조회 API 문서 추가

yesjuhee committed:
feat: 어드민 예약 대기 목록 조회 기능 추가

- 어드민 예약 대기 목록 조회를 위한 `AdminWaitingController` 및 관련 DTO 추가
- `WaitingService에` getWaitings() 메서드 추가
- 예약 대기 목록 조회 기능의 JS 연동 및 데이터 매핑 구현
- AdminWaitingApiTest를 통해 조회 API 정상 작동 검증

yesjuhee committed:
feat: 어드민 예약 대기 거절 기능 추가

- `DELETE /admin/waitings/{id}`` API 추가 및 컨트롤러, 서비스 로직 구현
- 예약 대기 거절 기능에 대한 프론트엔드 로직 보완
- 어드민 예약 대기 거절 테스트 및 예외 처리 케이스 추가
- 관련 API 문서 작성 및 업데이트

yesjuhee committed:
docs: 예약 자동 승인 기능관련 문서 추가

yesjuhee committed:
feat: 예약 취소 시 대기 목록 자동 업가데이트 기능 추가

- 예약 취소 시 대기 목록 조회 및 다음 대기자를 예약자로 설정하는 로직 추가
- `ReservationService`에 대기 목록 확인 및 업데이트 기능 구현
- `WaitingRepository`에 `findFirstByReservationIdOrderByCreatedAtAsc` 쿼리 메서드 추가
- 테스트 코드 추가해 예약 취소 및 대기 목록 업데이트 검증

yesjuhee committed:
docs: 어드민 예약 생성 API 문서 추가 및 API 문서 상세화

- 어드민 예약 생성 API 요청 및 응답 예시 추가
- DELETE 요청 경로를 {id} 형식으로 통일 (`/themes/1` → `/themes/{id}`)
- 시간 조회 API 제목 변경 (시간 조회 → 시간 목록 조회)
- 전체 API 리스트를 세부 경로와 함께 업데이트

yesjuhee committed:
docs: 어드민 예약 생성 API 문서 추가 및 API 문서 상세화

- 어드민 예약 생성 API 요청 및 응답 예시 추가
- DELETE 요청 경로를 {id} 형식으로 통일 (`/themes/1` → `/themes/{id}`)
- 시간 조회 API 제목 변경 (시간 조회 → 시간 목록 조회)
- 전체 API 리스트를 세부 경로와 함께 업데이트

yesjuhee committed:
refactor: `ReservationRepository` `findAll` 메서드에 Fetch Join 추가

yesjuhee committed:
chore: 주석 삭제

yesjuhee committed:
refactor: 사용되지 않는 테스트 비활성화

kang-hyungu commented:
@coderabbitai review

yesjuhee commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -22,13 +25,13 @@ public class Reservation {
+    @ManyToOne(fetch = FetchType.LAZY)
     private ReservationTime time;
 
-    @ManyToOne
+    @ManyToOne(fetch = FetchType.LAZY)
     private Member member;
-    
-    @ManyToOne
+
+    @ManyToOne(fetch = FetchType.LAZY)
     private Theme theme;
```
[지난번 리뷰](https://github.com/woowacourse/spring-roomescape-waiting/pull/216#discussion_r2094573717)를 반영해서 `@ManyToOne`의 연관관계 패치 옵션을 `LAZY`로 설정했습니다. 또한 `ResevationRepository::findAll`의 경우 n + 1 현상 방지를 위해 `fetch join`을 사용하도록 수정했습니다!

yesjuhee commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -42,22 +45,32 @@ public Reservation(final Long id, final Member member, final LocalDate date, fin
     }
 
-    protected Reservation() {
+    public void updateMember(final Member member) {
+        this.member = member;
+    }
```
### 더티체킹과 불변객체

예약 취소시 예약 대기 자동 승인 기능을 구현하면서 해당 메서드를 만들었습니다. 예약이 취소되면 해당 예약과 연관된 예약 대기를 검색해서, 예약 대기가 존재하면 대기하던 멤버를 예약의 멤버로 업데이트합니다.

JPA의 더티체킹을 사용하면 객체의 필드 값이 변경된 것을 자동 감지해 DB에 업데이트 쿼리를 날려주는 것으로 알고 있어서 해당 기능을 활용했습니다. (깊게 공부해보진 못했는데 이렇게 사용하는걸 본 기억이 있어서 사용했습니다!)

레벨1에서 미션을 할 때는 가능하면 객체를 불변으로 만들고 업데이트시 새로운 객체를 반환하는 방식을 사용했습니다. 그런데 JPA와 엔티티의 개념이 도입되면서 레벨1의 기존 방식과 충돌하는 것 같은 기분이 들었습니다.

엔티티는 속성이 변해도 식별자로서 구분되는 것이기 때문에 오히려 속성이 변경될 수 있는 가변객체인 것이 더 자연스럽다고 생각하는게 맞을까요? 레벨1에서 사용했던 불변객체는 값객체를 위해 사용한다고 생각하는게 좋을까욥?

Rok93 replied:
현재 애플리케이션에 `예약 취소` API는 있지만 실제로 취소하는 기능은 없는걸까요? 
예약 취소를 직접해서 해당 동작이 제대로 동작하는지 테스트해보고 싶었는데, 취소 기능을 찾지 못해서 테스트를 못해봤네요. 

이 `DELETE /reservations/{id}` API에 대한 인수테스트 혹은 `reservationService#cancelReservationById` 메서드에 대한 통합 테스트 (w `@SpringBootTest`)를 작성해보시고 정상적으로 동작하는지 확인해보시면 좋을 것 같습니다. 👀

더티 체킹이 동작하기 위해서는 어떠한 조건이 필요한데요. 
이번에 이 기능을 사용하려고하셨으니 관련해서 공부를 해보시면 좋을 것 같아요. 😃

Rok93 replied:
> 레벨1에서 미션을 할 때는 가능하면 객체를 불변으로 만들고 업데이트시 새로운 객체를 반환하는 방식을 사용했습니다. 그런데 JPA와 엔티티의 개념이 도입되면서 레벨1의 기존 방식과 충돌하는 것 같은 기분이 들었습니다.
엔티티는 속성이 변해도 식별자로서 구분되는 것이기 때문에 오히려 속성이 변경될 수 있는 가변객체인 것이 더 자연스럽다고 생각하는게 맞을까요? 레벨1에서 사용했던 불변객체는 값객체를 위해 사용한다고 생각하는게 좋을까욥?

엔티티는 `식별자`를 기준으로 동일성을 판단하기 때문에, 속성이 바뀌더라도 여전히 같은 엔티티로 간주됩니다.

**엔티티**는 `식별자`로 구분되기 때문에 속성은 바뀔 수 있다는 가정 하에 설계되어야 하므로 **가변 객체(mutable object)**로 만드는 것이 일반적이라 생각해요. 이와 반대로  **값 객체**는 `속성`으로 구분되기 때문에 불변 객체로 만드는 것을 권장하는 것이라 생각해요.

yesjuhee replied:
> 현재 애플리케이션에 예약 취소 API는 있지만 실제로 취소하는 기능은 없는걸까요?
예약 취소를 직접해서 해당 동작이 제대로 동작하는지 테스트해보고 싶었는데, 취소 기능을 찾지 못해서 테스트를 못해봤네요.
이 DELETE /reservations/{id} API에 대한 인수테스트 혹은 reservationService#cancelReservationById 메서드에 대한 통합 테스트 (w @SpringBootTest)를 작성해보시고 정상적으로 동작하는지 확인해보시면 좋을 것 같습니다. 👀

앗 제 설명이 부족했네요 😅 
예약 취소는 어드민 페이지에서만 가능합니다! 어드민 계정으로 로그인 한 후 `http://localhost:8080/admin/reservation`로 접근하여 예약을 취소할 수 있습니다.

```
어드민 계정
- email: admin@gmail.com
- password: 1234
```
테스트의 경우 `reservation.ReservationApiTest.DeleteReservationTest`에 작성되어 있습니다!

yesjuhee replied:
> 더티 체킹이 동작하기 위해서는 어떠한 조건이 필요한데요.
이번에 이 기능을 사용하려고하셨으니 관련해서 공부를 해보시면 좋을 것 같아요. 😃

관련되어서 공부를 진행했습니다! 더티체킹은 트랜잭션이 커밋되는 시점에 발생합니다.
지금의 서비스 코드에는 트랜잭션이 명시되지 않았지만 더티체킹이 정상적으로 작동하고 있습니다.
리포지터리의 `delete()` 메서드에는 `@Transactional`이 포함되어 있기 때문에 `delete()` 메서드에 의해 트랜잭션이 생성되었다가 닫히고 이 시점에 더티체킹이 작동됩니다.

아직 제 코드에는 `@Transacitonal`이 필요한 부분을 찾지 못해서 적용하지 않았습니다!
관련되서 좀 더 학습해볼 만 한 부분이 있을까욥? 😄

Rok93 replied:
오... 그렇군요  언뜻보고 트랜잭션 안에서 동작하지 않으니 더티 체킹이 수행되지 않겠구나...? 라고 생각했었는데, 
`delete()` 메서드에 달린  트랜잭션 애너테이션 때문에 더티체킹이 수행되는가보군요? 🤔

조금 예상과는 달리 동작하여 저도 잘 인지하지 못했던 것 같아요. 😃
아직도 의심이되지만 이리저리 테스트해봤는데 말씀하신 것처럼 잘 동작하는 듯해서, 별도 수정은 안해주셔도될 것 같습니다. 👍

yesjuhee commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -24,17 +28,17 @@ public class ReservationService {
-            final ReservationRepository reservationRepository,
-            final ReservationTimeRepository reservationTimeRepository,
-            final ThemeRepository themeRepository,
-            final MemberRepository memberRepository
-    ) {
+    public ReservationService(final ReservationRepository reservationRepository,
+                              final ReservationTimeRepository reservationTimeRepository,
+                              final ThemeRepository themeRepository,
+                              final MemberRepository memberRepository, final WaitingRepository waitingRepository) {
```
### ReservationService의 의존성

예약 대기 서비스의 기능이 추가될수록 핵심 도메인인 Reservation의 서비스인 `ReservationService`의 의존성이 점점 늘어나고 있습니다. 예약 관련 비즈니스 로직을 처리하려면 연관된 도메인에 대한 작업이 필요하기 때문입니다.

기존에는 크게 문제를 느끼지 못하다가, 이정도로 의존성이 늘어나니까 `ReservationService`가 너무 많은 책임을 지고 있다는 생각이 들었습니다 😢 인스턴스 변수를 많이 가지고 있다는건 **SRP 위반**의 신호로 느껴집니다.

그렇다고 뭔가 뾰족한 해결방법이 떠오르지도 않네요..ㅎ **서비스가 다른 서비스를 의존하도록 하는 방식**도 있을 것 같습니다. 그래도 결국 인스턴스 변수의 종류가 `Repository`에서 `Service`로 바뀔 뿐 인스턴스 변수는 줄어들지 않겠네요😅 그래도 현재 로직에서 각 레포지터리를 가져와서 처리하는 로직들을 서비스로 넘겨줄 수 있기 때문에 책임 분리가 지금보다 잘 이루어질 것 같습니다!

다만, 현재의 구조에서는 이게 괜찮은 구조인가 라는 의문이 듭니다. 지금의 서비스는 컨트롤러 계층과 소통하기 위해 요청/응답 형식에 맞춘 DTO를 사용하고 있습니다. 그래서 서비스끼리 의존하는 순간 요청/응답 형식 DTO로 소통을 하게되는데, **비즈니스 로직이 뷰에 결합되는 문제**가 생겨버립니다.

그렇다면 서비스끼리 의존하는게 아니라 서비스와 레포지터리 계층 사이에 **공용 서비스같은 계층(?)** 을 만드는 것이 좋을까 하는 생각까지 하게 되었습니다! 아직 엄두는 안나서 생각만 해봤습니다ㅎㅎㅎ 이에 대한 로키의 의견이 궁금해요!

Rok93 replied:
좋은 고민을 해주신 것 같아요! 👍💯 

말씀하신 것처럼 서비스가 성장하면 성장할수록 여러 기능들이 생기게될 것이고 Service 계층도 비대해지게되는데, 이런 현상을 두고 `Fat Service`라고 표현해요. 

결론은 객체지향 프로그래밍처럼 서비스도 가능한 하나의 역할만 수행하도록 분리해야 더 유지보수하기 좋은 객체가 된다는 관점에서는 비슷한 것 같아요. 

먼저 `서비스가 다른 서비스를 의존하도록 하는 방식`은 그렇게 좋은 방식은 아니라고 생각해요. 🤔
계층을 나눴다면 가능한 같은 계층 끼리는 서로 의존하지 않도록 만드는 것이 의존성 방향을 단방향으로 깔끔하게 만들 수 있다고 생각해요. 그래서 같은 서비스 계층이라면 서로 의존하지 않도록하는 것이 좋다고 생각합니다. 

하지만 서비스 로직을 다짜고짜 SRP로 쪼개는 것은 오히려 유지보수하기 더 힘든 코드가 될 수 있다고 생각해요. 😃
잘 생각해보면 지금 느끼는 불편함은 하나의 서비스에 의존성이 점점 커지고 클래스 자체가 커지기 시작해 코드를 파악하기가 어려워진다는 점 때문인데요.   이 서비스 로직을 적절히 분리만해줘도 지금 정도의 프로젝트 규모에서는 괜찮은 해결책이 될 수 있다고 생각해요. 

예를들면 저 같은 경우에는  Command(= `C,U,D` 작업)와 Query(= Only 'R'ead 작업) Service를 별도로 쪼개서 사용하는 것을 선호하는 편이에요.  초기에는 이렇게만 분리해도 꽤 클래스의 덩치도 줄일 수 있고 서비스를 적절한 성질(?) 별로 분리하여 코드를 추적하거나 유지보수하기에도 괜찮았었어요. 

하지만 해당 서비스가 핵심 도메인을 다루고있다면 이렇게 분리해도 여전히 유지보수가 너무 어려운 상황이라면 저는 서비스가 하나의 역할만 하도록 분리할 것 같습니다. 

예를들면 `ReservationReserveService`, `ReservationCancelService`와 같은 방식으로 서비스를 분리하여 구성을 할 것 같아요.

Rok93 replied:
> 그렇다면 서비스끼리 의존하는게 아니라 서비스와 레포지터리 계층 사이에 공용 서비스같은 계층(?) 을 만드는 것이 좋을까 하는 생각까지 하게 되었습니다! 아직 엄두는 안나서 생각만 해봤습니다ㅎㅎㅎ 이에 대한 로키의 의견이 궁금해요!

아마 이렇게 생각하신 이유는 도메인을 C,R,U,D 할 때는 비슷한 패턴(?)이 있고, 각 서비스 계층에서 이런 비슷한 패턴이 공통적으로 쓰이는 경우가 있다보니 이런 고민을 하신게 아닐까 싶어요. 🤔  
제가 추측한 상황이 맞다면, 말씀하신 것과 같은 계층을 두는 것도 괜찮은 방식이라고 생각해요. 👍👍 
다만 계층을 추가한다는 것은 시스템이 한층 복잡해진다는 의미이기 때문에 이런 계층을 도입하신다면 고민포인트가 많아지시지 않을까 싶네요. 😉

yesjuhee replied:
> 먼저 서비스가 다른 서비스를 의존하도록 하는 방식은 그렇게 좋은 방식은 아니라고 생각해요. 🤔
계층을 나눴다면 가능한 같은 계층 끼리는 서로 의존하지 않도록 만드는 것이 의존성 방향을 단방향으로 깔끔하게 만들 수 있다고 생각해요. 그래서 같은 서비스 계층이라면 서로 의존하지 않도록하는 것이 좋다고 생각합니다.

동감합니다 👍 

> 예를들면 저 같은 경우에는 Command(= C,U,D 작업)와 Query(= Only 'R'ead 작업) Service를 별도로 쪼개서 사용하는 것을 선호하는 편이에요. 초기에는 이렇게만 분리해도 꽤 클래스의 덩치도 줄일 수 있고 서비스를 적절한 성질(?) 별로 분리하여 코드를 추적하거나 유지보수하기에도 괜찮았었어요.

오 이런 방법도 있군요! 한 번 반영해보도록 하겠습니다 🫡 

> 말씀하신 것과 같은 계층을 두는 것도 괜찮은 방식이라고 생각해요. 다만 계층을 추가한다는 것은 시스템이 한층 복잡해진다는 의미이기 때문에 이런 계층을 도입하신다면 고민포인트가 많아지시지 않을까 싶네요. 😉

계층의 복잡도를 늘릴만한 문제는 아닌 것 같네요ㅎㅎ Command - Query로 리팩터링 해보겠습니다!

0995fe2b
e85ba0c4

yesjuhee commented on src/main/java/roomescape/waiting/controller/AdminWaitingController.java:
```diff
@@ -0,0 +1,34 @@
+import roomescape.waiting.dto.WaitingSimpleResponse;
+import roomescape.waiting.service.WaitingService;
+
+@RestController
+@RequestMapping("/admin/waitings")
+public class AdminWaitingController {
```
관리자 권한이 필요한 API들은 `/admin/**`으로 API 경로를 설정하고 별도의 컨트롤러로 분리했습니다. 인증/인가 기능 활용 + 관심사 분리의 효과가 있습니다.

yesjuhee commented on src/main/java/roomescape/waiting/domain/Waiting.java:
### 예약 대기 설계

[설계 과정](https://prolog.techcourse.co.kr/studylogs/4400)에 대해서는 프롤로그에 자세히 적어놨습니다! 프롤로그에도 적혀있지만 지금의 설계는 제가 생각한 최선의 설계는 아닙니다. 기존의 설계 방식에서 토론 후 생각이 바뀌었습니다. 그런데 그때는 이미 기능이 많이 구현된 후라서 시간을 아끼고 빨리 PR을 보내기 위해 리팩터링은 하지 않았습니다! 대신 프롤로그에 생각 흐름을 확실히 정리해서 지금 구조의 문제점과 개선점에 대해서 탐구를 했습니다.

### 설계의 방향
이렇게 설계에 대해서 깊게 고민하면서 **객체 중심 설계**에 대한 궁금증이 생겼습니다. 저는 이번 미션에서 설계를 하면서 자연스럽게 DB의 구조부터 먼저 생각하고, ERD 다이어그램부터 그렸습니다. 그런데 토론을 할 때는 아예 도메인 객체 설계를 먼저 고려하는 크루도 있었습니다.

프롤로그에도 적었지만 JPA가 도입되면서 객체만 잘 설계해도 알아서 DB 스키마를 생성해주기 때문에 **JPA의 도움을 받아 객체 중심 설계가 가능**하겠다는 것을 배웠습니다. 미션1에서 설계를 했던 방식처럼 객체를 먼저 설계하는 것이 가능했던 것이죠..!

하지만 저는 이번 미션을 하면서 객체 중심 설계를 시도해봤지만 실패했습니다. 익숙하지 않아서 그런지 머리에 잘 그려지지 않았습니다 😅 또 객체지향 중심 사고를 했을 때 비효율적인 DB 스키마가 탄생되지 않을까 하는 의문도 남아있습니다. 아직까지의 생각으로는 객체지향 중심 설계는 이상적인 이론으로 느껴지고 **현실적으로는 DB 중심 설계가 맞지 않을까** 생각이 듭니다.

이는 저의 아주 짧은 경험에 의한 생각이기 때문에.. 이에 대한 로키의 의견도 궁금해요! 현업에서 DB 중심 설계를 하지 않고 객체 중심 설계를 하기도 하나요?

Rok93 replied:
DB 중심 설계를 먼저하는 것이 지향점은 아니지만, DB 중심 설계를 아예 배제할 수는 없다고 생각해요. 🤔
다만 저희가 더 나은 코드 품질과 유지보수를위해 객체지향 프로그래밍이라는 패러다임을 활용하고있는만큼 최대한 객체지향적인 설계를 가져가는 것이 좋다고 생각해요. 그래서 시작은 객체지향적으로하되 요구사항을 파악하며 이런 설계로도 대응이될지...? 를 고민하면서 필요한 부분은 타협을해나가는 것이 좋다고 생각해요. 

잠시 DB 적인 생각은 내려놓고 먼저 객체중심으로 설계를 빠르게한 뒤, DB를 접목시키면서 발생할 수 있는 문제 포인트들을 파악하며 조금씩 수정해나가는 것이 좋다고 생각해요. 

> 현업에서 DB 중심 설계를 하지 않고 객체 중심 설계를 하기도 하나요?

안하시는 분들이 훨씬 많은 것 같아요. 😅
저도 객체 중심으로 설계를 하는 경우도 있지만, 안하는 경우도 있어요. 😋

yesjuhee replied:
> 다만 저희가 더 나은 코드 품질과 유지보수를위해 객체지향 프로그래밍이라는 패러다임을 활용하고있는만큼 최대한 객체지향적인 설계를 가져가는 것이 좋다고 생각해요. 그래서 시작은 객체지향적으로하되 요구사항을 파악하며 이런 설계로도 대응이될지...? 를 고민하면서 필요한 부분은 타협을해나가는 것이 좋다고 생각해요.

좋은 의견 감사합니다ㅎㅎ 다음 미션에서는 객체 설계를 우선으로 하는 연습을 해봐야겠네요 😄

yesjuhee commented on src/main/java/roomescape/waiting/service/WaitingService.java:
```diff
@@ -0,0 +1,74 @@
+    public void deleteWaitingById(final long waitingId, final LoginMember loginMember) {
+        final Waiting waiting = waitingRepository.findById(waitingId)
+                .orElseThrow(() -> new BadRequestException("id를 찾을 수 없습니다."));
+        final Member member = memberRepository.findById(loginMember.id())
+                .orElseThrow(() -> new BadRequestException("회원 정보를 찾을 수 없습니다."));
+        if (!waiting.hasOwner(member)) {
+            throw new ForbiddenException("본인의 예약대기만 삭제할 수있습니다.");
+        }
```
### 중복 인증

아래 토끼의 리뷰를 보고 질문이 하나 더 생겼습니다 🐇 
사실 `LoginMember`는 `MemberArgumentResolver`에 의해서 이미 인증을 마친 회원의 정보입니다. 그런데 이 정보를 가지고 다시 `findById()`로 멤버를 가져오면서 올바른 아이디인지 까지 확인을 하고 있습니다. 이미 검증된 멤버인데 통째로 다시 조회를 해오는게 불필요하게 느껴지는 것 같아요.

이런 경우 `findById()`로 `Member` 객체를 조회하지 않고 `id`값을 이용해서 필요한 로직을 다 처리하는게 좋을까요? 예약 대기의 소유자를 판단하는데는 `id`로 충분할 것 같습니다. 다만 한번 더 조회를 하면 확실하게 검증할 수 있다는 점이 장점인 것 같아요.

프레젠테이션 계층에서 인증 부분은 모두 마쳤다고 믿고, `id`만 가져다 쓰는 방식으로 리팩터링을 해보고 싶습니다. 혹시 제가 놓치고 있는게 있는지 질문드리고 싶습니다!

Rok93 replied:
두 번 조회하는 것이 불필요하다는 점에서는 동의합니다만, 
개인적인 생각에서는 서비스 로직을 Member 객체 조회 없이 수행하도록 변경하기보다는 토큰을 읽어서 LoginMember를 만드는 과정에서 굳이 멤버 테이블을 조회하는 것을 개선해보면 어떨까 싶어요.

지금의 경우에는 인증이 필요한(= 토큰이 필요한) API의 경우 매번 member 테이블을 조회하고있을텐데요. 
많은 유저가 사용하는 서비스라면 무시하지 못할 정도로 member 테이블에대한 접근이 빈번할 것으로 보여요. 🤔
최초 로그인 시점에 member 테이블을 조회하여 토큰을 생성해준 것이니 이후 토큰은 인증이 끝난 것이니 별도로 추가 검증하지 않아도 되지 않을까요? 

그리고 `인증의 시점`과 `서비스 로직의 실행 시점`은 시간차가 조금 나기 때문에, 인증 시점의 member와 서비스 로직 실행 시점의 member는 달라질 수 있다는 점도 우려되는 포인트인 것 같아요. 👀

yesjuhee replied:
> 개인적인 생각에서는 서비스 로직을 Member 객체 조회 없이 수행하도록 변경하기보다는 토큰을 읽어서 LoginMember를 만드는 과정에서 굳이 멤버 테이블을 조회하는 것을 개선해보면 어떨까 싶어요.

오 아주 좋은 관점인 것 같습니다! 

- 토큰을 발행했을 때 member 테이블을 조회하여 발행한 토큰이니까 `MemberArgumentResolver`에서는 토큰의 유효성만 검증한다.
- 인증의 시점과 서비스 로직의 실행 시점은 차이가 있기 때문에 서비스 로직에서는 직접 조회를 통해 멤버를 확인한다.

yesjuhee replied:
0969be11

말씀해주신대로 토큰을 검증할 때 Member 테이블을 읽지 않고 토큰 검증만 하도록 `AuthService:checkMember` 메서드를 수정했습니다. 그런데 이 방향으로 리팩터링을 하면서 추가적인 고민이 생겼습니다.

**토큰에 사용자 이름을 추가해도 될까요?**
기존에는 토큰에 사용자 id만 있었고, 인증을 진행할 때 member 테이블을 읽어서 `LoginMember` DTO에 사용자 아이디, 이름, 권한, 이메일 정보를 모두 넣어서 사용했습니다. 이번에 member 테이블을 읽지 않는 방향으로 리팩터링을 했기 때문에 `LoginMember` DTO를 만들 때 토큰에 있는 정보를 사용해야 합니다.

`/admin/**` 경로를 검증할 때도 member 테이블을 읽지 않고 토큰의 권한을 검증하기 위해서 토큰에 `Role`을 추가했습니다.

토큰에 `name`을 넣은 것은 `GET /login/check` API를 위해서 입니다(`AuthController::loginCheck`에서 확인할 수 있습니다.) 이 API는 로그인 한 유저의 정보를 화면에 표시하기 위한 API로 로그인 한 유저의 이름을 보내줘야 합니다. 이 API도 현재 프론트엔드의 모든 페이지에서 호출되고 있습니다. 토큰에 `name`을 넣지 않고 이 API가 호출될 때마다 member 테이블을 조회해서 `name` 정보를 넣어주게 되면 리팩터링한 의미가 없을 것 같아서 이 API에서도 db를 조회하지 않도록 토큰의 정보를 사용하도록 했습니다.

토큰에 이름 정보를 넣지 않고 메번 member 테이블을 읽어서 `GET /login/check`에 응답하기
vs
토큰에 이름 정보 넣기
이 둘 중 저는 후자를 선택했습니다! 이 방식도 괜찮은가요? 아니면 이름은 민감한 정보이기 때문에 토큰에 넣지 않고 메번 테이블을 조회하는게 좋을까요?

Rok93 replied:
정답은 없다고 생각하지만 적어도 민감정보는 토큰이 탈취당하는 경우에 치명적일 수 있기 때문에 넣지 않는 것이 당연하다고 생각해요. (ex. 비밀번호, 주민번호 etc...) 

먼저 이름 같은 경우에는 개인적인 생각으로는 민감정보라고 생각하지는 않아요. (동명이인이 있는 것처럼 이름 자체가 엄청 민감한 정보는 아니라고 생각)

이름 혹은 이메일 같은 경우 토큰에 넣을지 말지가 고민되는 값들이라고 생각하는데요.
지금처럼 이런 값들이 점점 토큰에 추가하고싶은 유혹들이 생길텐데요. 😃
그러다보면 점점 토큰에는 더 많은 정보들이 추가되게될거라 생각해요. 🤔

저라면 `전자의 방식`으로 **필요한 타이밍에 member의 정보를 가져오는 것이 좋다고 생각**해요. 
다만 지금의 경우 매번 이 `GET /login/check` API를 호출하기 때문에 문제가 발생하는 것 같아요. 

물론 지금은 백엔드의 작업으로만 해결을해야하는 상황이다보니 최대한 테이블 조회를 적게하는 방향을 고려하셨다고 생각해요. 😃
다만 함께 협업을 하는 상황(= 프론트 개발자와 협업하는 상황), 프론트에서 최초에만 이 정보를 조회하고 이후에는 로컬 저장소에 조회한 결과 값을 저장하여 재사용하도록 해달라고 요청해볼 것 같아요.

yesjuhee replied:
오 프론트와의 협업까지는 미처 생각해보지 못했는데 너무 좋은 관점이네요!
완전 납득이 됐습니다ㅎㅎ
자세한 설명 감사합니다 👍

yesjuhee commented on src/main/java/roomescape/waiting/service/WaitingService.java:
```diff
@@ -0,0 +1,74 @@
+            throw new ForbiddenException("본인의 예약대기만 삭제할 수있습니다.");
+        }
+        waitingRepository.delete(waiting);
+    }
```
회원 정보 조회는 이미 `MemberArgumentResolver`에서 완료됐기 때문에 사실 저 예외는 발생하지 않을 것 같아. 너는 어떻게 생각해?

yesjuhee commented:
.

yesjuhee commented:
### 테스트

테스트는 `RestAssured`를 이용한 통합테스트만 작성했습니다. 해당 테스트를 선택한 이유는 [프롤로그](https://prolog.techcourse.co.kr/studylogs/4400)를 참고해주세요 😄 

추가로 학습 키워드에 등장한 `@WebMvcTest`, `@DataJpaTest`를 이용해 슬라이스 테스트도 작성해보고 싶었으나 시간상 PR 제출 뒤에 시간이 되면 진행해 볼 예정입니다. 서비스 단위 테스트의 경우 기존에 Fake 객체를 사용해서 작성했었는데 미션을 진행하면서 생각이 바뀌었습니다. 리포지터리의 책임이 늘어나면서 Fake 객체를 구현하는 비용이 너무 많이 드는 것 같아 서비스 단위 테스트를 더 작성한다면 Fake 객체를 사용하지 않고 H2 DB까지 연결해서 단위 테스트를 진행할 것 같습니다.

Rok93 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -69,13 +73,29 @@ public List<ReservationResponse> getReservations(
     public void cancelReservationById(final long id) {
+        final Optional<Waiting> waiting = waitingRepository.findFirstByReservationIdOrderByCreatedAtAsc(id);
+        if (waiting.isPresent()) {
+            Waiting waiting2 = waiting.get();
```
(waiting2) 네이밍에 신경써보시면 좋을 것 같아요. 👀 

[Optional을 올바르게 쓰는 법](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/) 글을 한번 참고해보시면 좋을 것 같아요. 😃

yesjuhee replied:
앗 일단 쓰고 나중에 수정한다는걸 못고쳤네요 😅 
아래 커밋에서 해당 메서드 리팩터링을 진행했습니다!
`isPresent()` - `get()` 대신 `ifPresentOrElse()`를 사용해서 존재하는 경우와 존재하지 않는 경우의 행동을 각각 람다식으로 작성해주는 방햐으로 수정했습니다!

0995fe2b

Rok93 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -69,13 +73,29 @@ public List<ReservationResponse> getReservations(
     public void cancelReservationById(final long id) {
+        final Optional<Waiting> waiting = waitingRepository.findFirstByReservationIdOrderByCreatedAtAsc(id);
+        if (waiting.isPresent()) {
+            Waiting waiting2 = waiting.get();
```
존재하면 가져와서 처리하고 없으면 아무 예외없이 종료하는 방식은 정상적으로 처리가 된 것인지 아닌지 오해할 여지가 있다고 생각해요. 🤔

예약 대기를 조회했을 경우 존재하지 않으면 바로 예외를 던져주는 것은 어떨까요? 
```suggestion
        Waiting waiting = waitingRepository.findFirstByReservationIdOrderByCreatedAtAsc(id)
            .orElseThrow(() -> new BadRequestException("예약을 찾을 수 없습니다."));
```

yesjuhee replied:
이 로직은 아래와 같은 기능을 수행합니다.
- 삭제하고자 하는 예약에 **대기가 없는 경우** -> 예약을 삭제한다
- 삭제하고자 하는 예약게 **대기가 있는 경우** -> 기존 예약 정보를 대기 정보로 교체하고, 예약은 삭제하지 않는다
즉 대기가 없는 경우에도 예외가 발생하지 않아야 합니다!

대기가 예약을 참고하는 구조 때문에 메서드가 예상과 다르게 행동하여 혼란을 가져다 주었네요ㅎㅎ
리팩터링을 하면서 `ifPresentOrElse()` 메서드를 사용하여 혼동을 최소한으로 하였습니다

0995fe2b

Rok93 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -69,13 +73,29 @@ public List<ReservationResponse> getReservations(
+        if (waiting.isPresent()) {
+            Waiting waiting2 = waiting.get();
+            final Reservation reservation = reservationRepository.findById(id)
+                    .orElseThrow(() -> new BadRequestException("예약을 찾을 수 없습니다."));
```
리소스를 찾지 못한 케이스는 `BadRequestException` 예외를 던져주고있는데요. 
그리고 이 BadRequestException은 네이밍대로 전역 예외 핸들러테핸들러에서 400 응답을 내려주고있네요.  🤔

리소스를 찾지 못한 경우에는 400 보다 더 적절한 에러 응답이있지 않을까요? 😃

yesjuhee replied:
8ed68c22

404를 반환하는 `NotFoundException` 예외를 커스텀 예외로 추가했습니다!
이 예외는 요청 URI 경로에 쓰인 id가 잘못되었을 경우에만 작동하도록 설정했습니다!

Rok93 reviewed (changes requested):
노랑 3~4단계도 잘 구현해주셨네요. 👍👍 
질문 주신 내용에대해 제 생각 남겨두었으니 확인 부탁드릴게요. 😃
질문 주신 내용에대한 코멘트만 남겨도 꽤 고민해볼 포인트가 많은 것 같아서 테스트 관련해서는 다음 리뷰 때, 조금 더 집중적으로 보도록할게요. 😉

잘 이해가 안되거나 어려운 부분있으면 언제든지 DM 주세요. 🤗

yesjuhee committed:
refactor: 예약 서비스 Query와 Command 역할 분리

- `ReservationQueryService`와 `ReservationCommandService`를 각각 생성하여 예약 관련 로직 분리
- `cancelReservationById` 메서드 리팩터링

yesjuhee committed:
refactor: 예약 컨트롤러 분리

- `ReservationAdminController` 생성 및 관리자 전용 예약 관리 API 분리
- DELETE /reservations/{id} 엔드포인트를 /admin/reservations/{id}로 변경
- 관련 테스트 및 문서를 관리자 전용 API에 맞게 수정

yesjuhee committed:
refactor: `MemberArgumentResolver`에서 인증할 때 DB 조회 없이 토큰 검증만 하도록 리팩터링

- `JwtTokenProvider`에서 `Role`과 `name` 정보를 토큰에 포함하도록 수정
- `AuthService`에서 회원 식별 및 권한 확인 로직 단순화

yesjuhee committed:
refactor: equals 및 hashCode 메서드 개선

- `instanceof`와 패턴 매칭을 활용하여
- `id` 필드 호출 대신 `getId()` 메서드 호출
- `hashCode` 메서드의 구현을 `getId()` 기반으로 수정
- 프록시 객체의 `equals()` 사용을 고려하여 수정함

yesjuhee committed:
refactor: 예약대기 삭제 메서드 이름 변경

- `deleteWaitingById`를 `deleteMyWaiting` 및 `deleteWaitingByAdmin`으로 변경
- 메서드 명을 역할에 맞게 명확히 수정하여 가독성 및 코드 이해도 향상

yesjuhee committed:
refactor: 예약 대기 삭제 시 회원 조회 순서 수정

- 회원 정보를 먼저 조회하도록 로직 순서를 수정

yesjuhee committed:
refactor: `NotFoundException` 커스텀 예외 추가 및 예외 처리 방식 수정

- `NotFoundException`: 요청 URI의 리소스가 잘못되었음
- `UnauthorizedException`: 요청한 사람의 권한이 잘못되었음
- `BadRequestException`: 위 2개의 케이스를 제외한 남은 모든 상황

yesjuhee commented:
리뷰 내용 모두 반영했습니다 🫡 
시간이 없어서 추가 테스트는 작성하지 못했습니다 😂 
리뷰 감사합니다 👍

Rok93 commented on src/main/java/roomescape/waiting/controller/AdminWaitingController.java:
```diff
@@ -0,0 +1,34 @@
+        return ResponseEntity.ok().body(responses);
+    }
+
+    @DeleteMapping("/{id}")
```
예약 거절이 제대로 동작하지 않는 것 같은데, 확인해서 수정 부탁드릴게요. 😃

<img width="1800" alt="image" src="https://github.com/user-attachments/assets/38f63b54-88d6-4061-a372-c4fcdcbee514" />

yesjuhee replied:
b67d7a72

어이쿠 수정했습니다 🙇 🙇‍♂️

Rok93 commented on src/main/java/roomescape/waiting/controller/AdminWaitingController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> delete(@PathVariable("id") final long id) {
```
사소할 수 있지만 실제로 이 API는 예약을 거절하는 API 핸들러인데, 네이밍은 DB 관점으로 지으신 것 같네요. 👀

yesjuhee replied:
이 부분에 대해서 고민을 해봤는데 기존의 `delete`를 유지하는게 좋을 것 같다는 생각을 했습니다!
HTTP 메서드 delete와 연결지어서 이해하여 특정 리소스를 삭제하는 역할을 하는 API 컨트롤러임을 나타내기에 delete가 좋은 것 같습니다 😃 
이 부분을 생각하다가, 서비스 코드의 경우 비즈니스 용어를 사용하는 것이 가독성과 이해 측면에서 좋을 것 같아서 대신 서비스 메서드를 수정했습니다 👍 

67be7f14

Rok93 commented on src/main/java/roomescape/waiting/domain/Waiting.java:
```diff
@@ -0,0 +1,101 @@
+        this.member = member;
+    }
+
+    public Waiting() {
```
Entity로 지정하기위해서는 기본 생성자가 필요한데요. 😃
먼저 왜 기본 생성자가 필요한지 가볍게 공부해보시면 좋을 것 같아요. (동작 원리까지 자세히 살펴보라는 이야기는 아니에요!)

추가적으로 사실 기본생성자를 굳이 공개하고싶지 않은데 JPA의 제약사항에의해 억지로 열어둔 상태인데요. 🥲
하다못해 접근 제한자를 `protected`로 제한해보면 어떨까요? 
기존처럼 해당 객체를 Entity로 잘 인식하는지도 애플리케이션 구동해보면서 확인해보시면 좋을 것 같아요. 😉

yesjuhee replied:
앗 [저번 단계](https://github.com/woowacourse/spring-roomescape-waiting/pull/216#discussion_r2094150920)에서 학습을 했는데 그새 까먹고 퍼블릭을 썼네요... 😅 
꼼꼼한 확인 감사합니다 👍 

332c7162

Rok93 replied:
제가 1~2단계에서 분명 얘기했을 것 같았는데, 혹시 빼먹었나싶어서 다시 코멘트 남겼는데, 다행히도 잘 남겼었나보군요! 😉

Rok93 commented on src/main/java/roomescape/waiting/domain/Waiting.java:
```diff
@@ -0,0 +1,101 @@
+        this.createdAt = createdAt;
+        this.reservation = reservation;
+        this.member = member;
+    }
```
사실 id를 직접 지정하실 일은 없을 것 같은데, 생성자에서도 굳이 id를 주입받을 필요 없지 않을까요? 🤔

yesjuhee replied:
지금까지 AllArgsConstructor를 생성해두면 유용하게 사용했어서 습관처럼 정의했었네요. 다시 생각해보니 JPA를 도입하면서 ID를 생성자로 받는 케이스는 거의 없겠군요! 반영했습니다 👍 

332c7162

Rok93 reviewed (changes requested):
노랑 마지막까지 피드백 반영 잘해주셨어요 👍👍
현재 동작하지 않는 곳이 있어서, 이 부분만 빠르게 수정부탁드릴게요. 🙏
다음 리뷰 요청 때는 방탈출 예약 대기 미션은 이만 종료하도록하겠습니다. 😉

yesjuhee committed:
refactor: API 요청 형식 변경사항 반영 및 사용하지 않는 스크립트 삭제

yesjuhee committed:
refactor: 서비스 메서드 명을 비즈니스 도메인의 성격을 띄도록 변경

yesjuhee committed:
refactor: `Waiting` 생성자 및 메서드 수정

yesjuhee commented:
마지막까지 꼼꼼한 확인 감사드립니다!! 👍 🚀

Rok93 reviewed (approved):
노랑 마지막까지 피드백 반영 잘해주셨어요. 👍👍 
방탈출 예약 대기 미션은 이만 종료하도록할게요. 🎉👏
Lv2 마지막 미션도 화이팅입니다. 💪💪
